### PR TITLE
[staging] llvmPackages*.libllvm: drop outputs.patch for llvm-config.patch

### LINKED ISSUE #127505: llvm-config-native --libs --link-shared fails because it can't find libLLVM-9.so
**Describe the bug**

```console
$ nix-build -A pkgsCross.aarch64-multiplatform.llvmPackages_9.llvm.dev
$ ./result-dev/bin/llvm-config-native --libs --link-shared
llvm-config: error: libLLVM-9.so is missing
```

**To Reproduce**

More specifically I was trying to get `haskellPackages.llvm-hs` to cross compile (which may or may not be possible).
This is what I have so far which seems to work, but it fails doing `configurePhase` with the above error. Adding
`llvmPackages_9.llvm.lib` (or `llvmPackages_9.libllvm`) to `librarySystemDepends` or `libraryPkgconfigDepends`
does not help.

Note: The patch should be applied on top of  #127494 to get the dependencies of `llvm-hs` to cross-compile.

```diff
diff --git a/pkgs/development/haskell-modules/configuration-nix.nix b/pkgs/development/haskell-modules/configuration-nix.nix
index 161712d594b..649be69c69a 100644
--- a/pkgs/development/haskell-modules/configuration-nix.nix
+++ b/pkgs/development/haskell-modules/configuration-nix.nix
@@ -240,7 +240,19 @@ self: super: builtins.intersectAttrs super {
   caramia = dontCheck super.caramia;
 
   # requires llvm 9 specifically https://github.com/llvm-hs/llvm-hs/#building-from-source
-  llvm-hs = super.llvm-hs.override { llvm-config = pkgs.llvm_9; };
+  llvm-hs = overrideCabal (super.llvm-hs.override {
+    # llvm-config should not come from buildPackages since it needs
+    # to tell us something about LLVM for the host platform. Luckily
+    # the llvm package contains llvm-config-native for this job.
+    llvm-config = pkgs.lib.getDev pkgs.llvmPackages_9.llvm;
+  }) (drv: {
+    # For debugging purposes
+    configureFlags = [ "-v3" ] ++ drv.configureFlags or [];
+    # Use llvm-config-native when we are cross compiling
+    postPatch = pkgs.lib.optionalString (with pkgs.stdenv; hostPlatform != buildPlatform) ''
+      substituteInPlace Setup.hs --replace llvm-config llvm-config-native
+    '' + (drv.postPatch or "");
+  });
 
   # Needs help finding LLVM.
   spaceprobe = addBuildTool super.spaceprobe self.llvmPackages.llvm;
```

<details>
<summary>Build log:</summary>

```
lukas@wolfgang ~/s/n/nixpkgs (haskell-cross-fixes)> nix-build -A pkgsCross.aarch64-multiplatform.haskellPackages.llvm-hs
these derivations will be built:
  /nix/store/bc8nmnn87pf0fjiwrdsbm5rb262svxra-llvm-hs-9.0.1-aarch64-unknown-linux-gnu.drv
building '/nix/store/bc8nmnn87pf0fjiwrdsbm5rb262svxra-llvm-hs-9.0.1-aarch64-unknown-linux-gnu.drv'...
setupCompilerEnvironmentPhase
Build with /nix/store/g4ylg7rmfikskbg9iryc1i8zcik7paq6-aarch64-unknown-linux-gnu-ghc-8.10.4.
unpacking sources
unpacking source archive /nix/store/2xmdp36vnlr5kpxw3wi4lpcsjc6qma77-llvm-hs-9.0.1.tar.gz
source root is llvm-hs-9.0.1
setting SOURCE_DATE_EPOCH to timestamp 1000000000 of file llvm-hs-9.0.1/test/Test.hs
patching sources
compileBuildDriverPhase
setupCompileFlags: -package-db=/build/setup-package.conf.d -j4 +RTS -A64M -RTS -threaded -rtsopts
[1 of 1] Compiling Main             ( Setup.hs, /build/Main.o )
Linking Setup ...
updateAutotoolsGnuConfigScriptsPhase
configuring
configureFlags: --verbose --prefix=/nix/store/d5fdjy2xlk1dj6v1mhnbyjvd9larynvx-llvm-hs-9.0.1-aarch64-unknown-linux-gnu --libdir=$prefix/lib/$compiler --libsubdir=$abi/$libname --docdir=/nix/store/kvwp4hd2cwr00bmq1547x7gi48b8q41v-llvm-hs-9.0.1-aarch64-unknown-linux-gnu-doc/share/doc/llvm-hs-9.0.1 --with-gcc=aarch64-unknown-linux-gnu-gcc --package-db=/build/package.conf.d --ghc-options=-j4 +RTS -A64M -RTS --disable-split-objs --disable-library-profiling --disable-profiling --enable-shared --disable-coverage --enable-static --disable-executable-dynamic --disable-tests --disable-benchmarks --enable-library-vanilla --disable-library-for-ghci --ghc-option=-split-sections --configure-option=--host=aarch64-unknown-linux-gnu --with-ghc=aarch64-unknown-linux-gnu-ghc --with-ghc-pkg=aarch64-unknown-linux-gnu-ghc-pkg --with-gcc=aarch64-unknown-linux-gnu-cc --with-ld=aarch64-unknown-linux-gnu-ld --with-ar=aarch64-unknown-linux-gnu-ar --with-hsc2hs=aarch64-unknown-linux-gnu-hsc2hs --with-strip=aarch64-unknown-linux-gnu-strip --hsc2hs-option=--cross-compile -v3 --extra-lib-dirs=/nix/store/hqp6jjj8rcb34cvmv048kn7nxfsnj3xp-ncurses-6.2-aarch64-unknown-linux-gnu/lib --extra-lib-dirs=/nix/store/d4x417p5phnf95xja1wl2v6jmc179c6y-libffi-3.3-aarch64-unknown-linux-gnu/lib --extra-lib-dirs=/nix/store/qbip7mr9r4cz6vh3bs5gsr6xc6d30hiq-gmp-6.2.1-aarch64-unknown-linux-gnu/lib
Using Parsec parser
Searching for llvm-config-native-9.0 in path.
Cannot find llvm-config-native-9.0 on the path
Searching for llvm-config-native-9 in path.
Cannot find llvm-config-native-9 on the path
Searching for llvm-config-native in path.
Found llvm-config-native at
/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native
/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native --version
/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native
is version 9.0.1
/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native --cxxflags
/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native --includedir
/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native --libdir
/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native --version
/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native --libs --link-shared
/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native
returned ExitFailure 1 with error message:
llvm-config: error: libLLVM-9.so is missing
CallStack (from HasCallStack):
  $, called at libraries/Cabal/Cabal/Distribution/Simple/Program/Run.hs:160:7 in Cabal-3.2.1.0:Distribution.Simple.Program.Run
  getProgramInvocationOutput, called at libraries/Cabal/Cabal/Distribution/Simple/Program.hs:152:3 in Cabal-3.2.1.0:Distribution.Simple.Program
  getProgramOutput, called at Setup.hs:80:12 in main:Main
Setup:
'/nix/store/6kky4f7ncgrff79skd2lqmgah6vwqyk4-llvm-aarch64-unknown-linux-gnu-9.0.1-dev/bin/llvm-config-native'
exited with an error:
llvm-config: error: libLLVM-9.so is missing

builder for '/nix/store/bc8nmnn87pf0fjiwrdsbm5rb262svxra-llvm-hs-9.0.1-aarch64-unknown-linux-gnu.drv' failed with exit code 1
error: build of '/nix/store/bc8nmnn87pf0fjiwrdsbm5rb262svxra-llvm-hs-9.0.1-aarch64-unknown-linux-gnu.drv' failed
```

</details>

cc @Ericson2314

### PR DESCRIPTION
**Pull request by sternenseemann**
Due to gnu-install-dirs.patch llvm-config will return correct results
for --link-shared as well as --link-static even without this patch.

I am currently verifying this using the following script:

```bash
#!/bin/sh

versions=("5" "6" "7" "8" "9" "10" "11" "12" "13") # "git")

flags=(
  "--version"
  "--prefix"
  "--src-root"
  "--obj-root"
  "--bindir"
  "--includedir"
  "--libdir --link-shared"
  "--libdir --link-static"
  "--cmakedir"
  "--cppflags"
  "--cflags"
  "--cxxflags"
  "--ldflags"
  "--system-libs --link-shared"
  "--system-libs --link-static"
  "--libs --link-shared"
  "--libs --link-static"
  "--libnames --link-shared"
  "--libnames --link-static"
  "--libfiles --link-shared"
  "--libfiles --link-static"
  "--components"
  "--targets-built"
  "--host-target"
  "--build-mode"
  "--assertion-mode"
  "--build-system"
  "--has-rtti"
  "--has-global-isel"
  "--shared-mode"
)

outputs=("out" "dev" "lib")

for version in "${versions[@]}"; do
    echo "=== llvmPackages_$version ==="
    llvmConfig="$(nix-build -A llvmPackages_$version.libllvm.dev)/bin/llvm-config"

    transformOutputs=()

    for output in "${outputs[@]}"; do
        transformOutputs+=(
            "-e" "s|$(nix-build -A llvmPackages_$version.libllvm.$output)|$output|g"
        )
    done

    for flag in "${flags[@]}"; do
        "$llvmConfig" $flag | sed "${transformOutputs[@]}"
    done
done
```

Will take some build time, currently checked:

- [x] LLVM 5
- [x] LLVM 6
- [x] LLVM 7
- [x] LLVM 8
- [x] LLVM 9
- [x] LLVM 10
- [x] LLVM 11
- [x] LLVM 12
- [x] LLVM 13
- [ ] LLVM HEAD

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
Note that this *will* cause a regression for cross compilation (see also <https://github.com/NixOS/nixpkgs/pull/148367>), which can be (better) fixed along the lines of <https://github.com/NixOS/nixpkgs/pull/152776/commits/46240c36687ed7699628a7a97d360d0c91007d76>.

Basically `outputs.patch` will set `ActiveLibDir` to a hard coded `$lib/lib` value (previously only if `--link-shared`, since #148367 in both cases). The problem appears only in the first place because `llvm-config-native` is produced by running cmake configure again with a different host platform. For this, basically all flags passed to cmake originally are dropped, so we need to pass `CMAKE_INSTALL_*` again explicitly to the native configure step in order for `llvm-config-native` to produce correct results.

cc @r-burns

**Comment by sternenseemann** [comment]
The script pasted above produces the same output for normal `llvm-config` with and without this patch, which is great. I'll fix the cross situation tomorrow.

File: pkgs/development/compilers/llvm/10/llvm/default.nix [hunk 1]
```diff
@@ -167,7 +162,18 @@ in stdenv.mkDerivation (rec {
...
-      in "-DCROSS_TOOLCHAIN_FLAGS_NATIVE:list=${lib.concatStringsSep ";" nativeToolchainFlags}"
+        # We need to repass the custom GNUInstallDirs values, otherwise CMake
+        # will choose them for us, leading to wrong results in llvm-config-native
+        nativeInstallFlags = [
+          "-DCMAKE_INSTALL_PREFIX=${placeholder "out"}"
+          "-DCMAKE_INSTALL_BINDIR=${placeholder "bin"}/bin"
+          "-DCMAKE_INSTALL_INCLUDEDIR=${placeholder "dev"}/include"
+          "-DCMAKE_INSTALL_LIBDIR=${placeholder "lib"}/lib"
+          "-DCMAKE_INSTALL_LIBEXECDIR=${placeholder "lib"}/libexec"
+          "-DLLVM_INSTALL_CMAKE_DIR=${placeholder "dev"}/lib/cmake/llvm/"
```
Comment (Ericson2314) [inline pkgs/development/compilers/llvm/10/llvm/default.nix:173]: I will aim to make this one not necessary to fix soon :)

**Review by Ericson2314 — APPROVED** [review]

**Comment by sternenseemann** [comment]
Okay, so now `--link-shared` is broken for cross, but `--link-static` works… I'll need to investigate what's going on.

**Comment by sternenseemann** [comment]
Okay, I think I've finally untangled it. As it turns out `llvm-config` uses `LLVM_ENABLE_DYLIB` and `LLVM_LINK_DYLIB` to determine whether a) `libLLVM.so` is available and b) whether it should be used by default (`--shared-mode` uses this for example).

Both these values are influenced by `LLVM_LINK_LLVM_DYLIB=ON` in our case, so we need to pass this via `CROSS_TOOLCHAIN_FLAGS_NATIVE` as well — except that does not work because building the full (!) `libLLVM.so` as part of `NATIVE` just fails (and is also undesireable). To work around this we need to patch `llvm-config` to link statically against the relevant LLVM components.

As you can see for LLVM 11, the `BuildVariables.inc` used by `llvm-config` is now practically equivalent (I'll leave the system libs part as an exercise for later…). What's left now is to check all the builds (again), i. e. I don't know yet if the patch applies everywhere.

```diff
--- build/NATIVE/tools/llvm-config/BuildVariables.inc	2022-01-03 13:06:13.482576812 +0100
+++ build/tools/llvm-config/BuildVariables.inc	2022-01-03 13:05:36.917470071 +0100
@@ -16,19 +16,19 @@
 //===----------------------------------------------------------------------===//
 
 #define LLVM_SRC_ROOT "/home/lukas/src/nix/nixpkgs/llvm-11.1.0.src"
-#define LLVM_OBJ_ROOT "/home/lukas/src/nix/nixpkgs/llvm-11.1.0.src/build/NATIVE"
+#define LLVM_OBJ_ROOT "/home/lukas/src/nix/nixpkgs/llvm-11.1.0.src/build"
 #define LLVM_CPPFLAGS "-D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS"
 #define LLVM_CFLAGS " -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS"
 #define LLVM_LDFLAGS ""
 #define LLVM_CXXFLAGS "-std=c++14   -fno-exceptions -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS"
 #define LLVM_BUILDMODE "Release"
 #define LLVM_LIBDIR_SUFFIX ""
-#define LLVM_INSTALL_BINDIR "/1rz4g4znpzjwh1xymhjpm42vipw92pr73vdgl6xs1hycac8kf2n9/bin"
-#define LLVM_INSTALL_LIBDIR "/0sra2y18lr3h6j58qjm0w46yv36h1wjmilb09n8aimdpivdymscx/lib"
-#define LLVM_INSTALL_INCLUDEDIR "/02qcpld1y6xhs5gz9bchpxaw0xdhmsp5dv88lh25r2ss44kh8dxz/include"
+#define LLVM_INSTALL_BINDIR "/nix/store/f8w4zhgy974b3hsvfif3b6fj4wfb5506-llvm-aarch64-unknown-linux-gnu-11.1.0/bin"
+#define LLVM_INSTALL_LIBDIR "/nix/store/609l10l12vdlgq5vc72sfi4iaiqnv5bh-llvm-aarch64-unknown-linux-gnu-11.1.0-lib/lib"
+#define LLVM_INSTALL_INCLUDEDIR "/nix/store/2ajndk42qqh609ml69lih3liwigvwc43-llvm-aarch64-unknown-linux-gnu-11.1.0-dev/include"
 #define LLVM_INSTALL_CMAKEDIR "/02qcpld1y6xhs5gz9bchpxaw0xdhmsp5dv88lh25r2ss44kh8dxz/lib/cmake/llvm/"
 #define LLVM_TARGETS_BUILT "AArch64 AMDGPU ARM AVR BPF Hexagon Lanai Mips MSP430 NVPTX PowerPC RISCV Sparc SystemZ WebAssembly X86 XCore"
-#define LLVM_SYSTEM_LIBS "-lrt -ldl -lpthread -lm -lxml2"
+#define LLVM_SYSTEM_LIBS "-lz -lrt -ldl -ltinfo -lpthread -lm -lxml2"
 #define LLVM_BUILD_SYSTEM "cmake"
 #define LLVM_HAS_RTTI 1
 #define LLVM_ENABLE_DYLIB 1
@@ -37,4 +37,4 @@
 #define LLVM_DYLIB_COMPONENTS "all"
 #define LLVM_DYLIB_VERSION "11"
 #define LLVM_HAS_GLOBAL_ISEL 1
-#define LLVM_TOOLS_INSTALL_DIR "/1rz4g4znpzjwh1xymhjpm42vipw92pr73vdgl6xs1hycac8kf2n9/bin"
+#define LLVM_TOOLS_INSTALL_DIR "/nix/store/f8w4zhgy974b3hsvfif3b6fj4wfb5506-llvm-aarch64-unknown-linux-gnu-11.1.0/bin"
```

**Review by Ericson2314 — APPROVED** [review]
Great work!

**Comment by sternenseemann** [comment]
Okay, I've tested all LLVM versions again: `llvmPackages_*.llvm` `llvm-config`'s output hasn't changed to before this PR, `pkgsCross.*.llvmPackages_*.llvm`'s `llvm-config-native`'s output is identical to the former, except for the host LLVM triple (naturally) and the list of system libs. Since the missing system lib is always `-ltinfo`, I suspect this is related to the `ncurses` cross issue we have been working around and could be addressed with that in the future.

LLVM 5 has discrepancies in the output, listing less components than the native build, but I suppose this is a genuine LLVM cross compilation issue only present in the oldest LLVM version we ship, so I don't think it's worth fixing for now.

**Pull request merged**